File: omniwordlist/generator.py
"""
Core generation engine

Streaming combinator pipeline with checkpoint support and deduplication
"""

import itertools
import random
from typing import Iterator, List, Set, Optional
from pathlib import Path
import hashlib
from .config import Config
from .charset import expand_pattern, get_charset, CHARSET_LOWERCASE
from .transforms import apply_transforms
from .filters import create_filter_pipeline
from .error import GeneratorError


class Generator:
    """Main wordlist generator"""
    
    def __init__(self, config: Config):
        """
        Initialize generator
        
        Args:
            config: Configuration object
        """
        config.validate()
        self.config = config
        self.tokens_generated = 0
        self.dedup_hashes: Set[str] = set()
        
        # Initialize random seed if specified
        if config.seed is not None:
            random.seed(config.seed)
        
        # Create filter pipeline
        self.filter_pipeline = create_filter_pipeline(config.filters)
    
    def generate(self) -> Iterator[str]:
        """
        Generate tokens based on configuration
        
        Yields:
            Generated tokens
        """
        # Determine generation mode
        if self.config.pattern:
            yield from self._generate_pattern()
        elif self.config.enabled_fields:
            yield from self._generate_fields()
        else:
            yield from self._generate_charset()
    
    def _generate_charset(self) -> Iterator[str]:
        """Generate tokens from charset"""
        charset = self._resolve_charset()
        
        for length in range(self.config.min_length, self.config.max_length + 1):
            if self.config.permutations_only:
                # Generate permutations (no repeating characters)
                for combo in itertools.permutations(charset, length):
                    token = ''.join(combo)
                    processed_token = self._process_token(token)
                    if processed_token is not None:
                        yield processed_token
            else:
                # Generate combinations with replacement
                for combo in itertools.product(charset, repeat=length):
                    token = ''.join(combo)
                    processed_token = self._process_token(token)
                    if processed_token is not None:
                        yield processed_token
    
    def _generate_pattern(self) -> Iterator[str]:
        """Generate tokens using pattern matching (Crunch-style)"""
        pattern = self.config.pattern
        if not pattern:
            raise GeneratorError("No pattern specified")
        
        # Expand pattern to charset
        charset = expand_pattern(pattern, self.config.literal_chars)
        
        # Generate combinations based on pattern length
        length = len(pattern)
        for combo in itertools.product(charset, repeat=length):
            token = ''.join(combo)
            processed_token = self._process_token(token)
            if processed_token is not None:
                yield processed_token
    
    def _generate_fields(self) -> Iterator[str]:
        """Generate tokens using field-based approach"""
        from .fields import FieldManager
        
        if not self.config.enabled_fields:
            raise GeneratorError("No fields enabled")
        
        # Get field values
        field_values = []
        for field_id in self.config.enabled_fields:
            field = FieldManager.get_field(field_id)
            if field:
                field_values.append(field['examples'])
            else:
                # Use field_id as fallback
                field_values.append([field_id])
        
        # Generate combinations of field values
        for combo in itertools.product(*field_values):
            # Join with separator if specified, otherwise concatenate
            if self.config.separator:
                token = self.config.separator.join(combo)
            else:
                token = ''.join(combo)
            
            processed_token = self._process_token(token)
            if processed_token is not None:
                yield processed_token
    
    def _resolve_charset(self) -> str:
        """Resolve charset from configuration"""
        if self.config.charset:
            # Check if it's a named charset (specific known names only)
            named_charsets = ['lowercase', 'uppercase', 'digits', 'symbols', 
                            'hex-lower', 'hex-upper', 'alphanumeric', 'all']
            if self.config.charset in named_charsets:
                return get_charset(self.config.charset)
            # Otherwise treat as custom charset
            return self.config.charset
        
        # Default to lowercase
        return CHARSET_LOWERCASE
    
    def _process_token(self, token: str) -> Optional[str]:
        """
        Process and validate token
        
        Args:
            token: Token to process
            
        Returns:
            Processed token or None if should be filtered
        """
        # Apply prefix/suffix
        if self.config.prefix:
            token = self.config.prefix + token
        if self.config.suffix:
            token = token + self.config.suffix
        
        # Apply transforms
        if self.config.transforms:
            token = apply_transforms(token, self.config.transforms)
        
        # Check filters
        if not self.filter_pipeline.should_include(token):
            return None
        
        # Check start/end boundaries
        if self.config.start_string and token < self.config.start_string:
            return None
        if self.config.end_string and token > self.config.end_string:
            return None
        
        # Deduplication
        if self.config.dedupe:
            token_hash = hashlib.blake2b(token.encode()).hexdigest()
            if token_hash in self.dedup_hashes:
                return None
            self.dedup_hashes.add(token_hash)
        
        # Check limits
        if self.config.max_lines and self.tokens_generated >= self.config.max_lines:
            return None
        
        self.tokens_generated += 1
        return token
    
    def generate_list(self, limit: Optional[int] = None) -> List[str]:
        """
        Generate tokens as a list
        
        Args:
            limit: Optional limit on number of tokens
            
        Returns:
            List of generated tokens
        """
        tokens = []
        for i, token in enumerate(self.generate()):
            tokens.append(token)
            if limit and i + 1 >= limit:
                break
        return tokens
    
    def preview(self, sample_size: int = 10) -> List[str]:
        """
        Generate a preview of tokens
        
        Args:
            sample_size: Number of tokens to preview
            
        Returns:
            List of sample tokens
        """
        return self.generate_list(limit=sample_size)
    
    def estimate_count(self) -> int:
        """
        Estimate total number of tokens to be generated
        
        Returns:
            Estimated token count
        """
        if self.config.max_lines:
            return self.config.max_lines
        
        charset = self._resolve_charset()
        charset_size = len(set(charset))
        
        if self.config.permutations_only:
            # Permutations: P(n, r) = n! / (n-r)!
            total = 0
            for length in range(self.config.min_length, self.config.max_length + 1):
                if length <= charset_size:
                    # Calculate permutation count
                    perm = 1
                    for i in range(length):
                        perm *= (charset_size - i)
                    total += perm
            return total
        else:
            # Combinations with replacement: n^r
            total = 0
            for length in range(self.config.min_length, self.config.max_length + 1):
                total += charset_size ** length
            return total
    
    def get_stats(self) -> dict:
        """
        Get generation statistics
        
        Returns:
            Dictionary of statistics
        """
        return {
            'tokens_generated': self.tokens_generated,
            'estimated_total': self.estimate_count(),
            'dedup_cache_size': len(self.dedup_hashes),
            'config': self.config.to_dict(),
        }

File: src/generator.rs
/// Core generation engine
/// 
/// Streaming combinator pipeline with checkpoint support and deduplication

use std::sync::Arc;
use parking_lot::Mutex;
use blake2::Blake2b512;
use std::hash::Hasher;
use std::collections::HashSet;

pub struct Generator {
    config: crate::Config,
    state: Arc<Mutex<GeneratorState>>,
}

#[derive(Debug, Clone)]
struct GeneratorState {
    tokens_generated: u64,
    checkpoint: Option<String>,
    dedup_hashes: HashSet<u64>,
}

impl Generator {
    pub fn new(config: crate::Config) -> crate::Result<Self> {
        config.validate()?;
        
        Ok(Self {
            config,
            state: Arc::new(Mutex::new(GeneratorState {
                tokens_generated: 0,
                checkpoint: None,
                dedup_hashes: HashSet::new(),
            })),
        })
    }

    /// Generate tokens from charset
    pub fn generate_charset(&self) -> crate::Result<Vec<String>> {
        let charset = self.resolve_charset()?;
        self.generate_from_charset(&charset)
    }

    /// Generate tokens using pattern matching (Crunch-style)
    pub fn generate_pattern(&self) -> crate::Result<Vec<String>> {
        let pattern = self.config.pattern.as_ref()
            .ok_or_else(|| crate::Error::ConfigError("No pattern specified".to_string()))?;
        
        let literal_chars = self.config.literal_chars.as_deref();
        let charset = crate::charset::expand_pattern(pattern, literal_chars)?;
        
        self.generate_from_charset(&charset)
    }

    /// Generate tokens using field-based approach
    pub fn generate_fields(&self) -> crate::Result<Vec<String>> {
        if self.config.enabled_fields.is_empty() {
            return Err(crate::Error::ConfigError(
                "No fields enabled".to_string(),
            ));
        }

        let mut tokens = vec![String::new()];

        for field_id in &self.config.enabled_fields {
            if let Some(field) = crate::fields::FieldManager::get_field(field_id) {
                let field_tokens = field.examples.clone();
                let mut new_tokens = Vec::new();
                
                for existing in &tokens {
                    for field_token in &field_tokens {
                        new_tokens.push(format!("{}{}", existing, field_token));
                    }
                }
                
                tokens = new_tokens;
            }
        }

        Ok(tokens)
    }

    /// Generate all combinations up to max_length
    fn generate_from_charset(&self, charset: &str) -> crate::Result<Vec<String>> {
        let mut tokens = Vec::new();
        let start = self.config.start_string.clone();
        let end = self.config.end_string.clone();

        for len in self.config.min_length..=self.config.max_length {
            let len_tokens = self.generate_combinations(charset, len, start.clone(), end.clone())?;
            tokens.extend(len_tokens);
            
            if let Some(limit) = self.config.max_lines {
                if tokens.len() >= limit as usize {
                    tokens.truncate(limit as usize);
                    break;
                }
            }
        }

        Ok(tokens)
    }

    /// Generate combinations of given length
    fn generate_combinations(
        &self,
        charset: &str,
        length: usize,
        start: Option<String>,
        end: Option<String>,
    ) -> crate::Result<Vec<String>> {
        let chars: Vec<char> = charset.chars().collect();
        let mut tokens = Vec::new();
        
        if self.config.permutations_only {
            self.generate_permutations(&chars, length, &mut tokens)?;
        } else {
            self.generate_combinations_recursive(
                &chars,
                length,
                &mut String::new(),
                &mut tokens,
            )?;
        }

        // Filter by start/end if specified
        if let Some(start_str) = start {
            tokens = tokens.into_iter()
                .skip_while(|t| t < &start_str)
                .collect();
        }

        if let Some(end_str) = end {
            tokens = tokens.into_iter()
                .take_while(|t| t <= &end_str)
                .collect();
        }

        // Apply prefix/suffix
        if let Some(prefix) = &self.config.prefix {
            tokens = tokens.into_iter()
                .map(|t| format!("{}{}", prefix, t))
                .collect();
        }

        if let Some(suffix) = &self.config.suffix {
            tokens = tokens.into_iter()
                .map(|t| format!("{}{}", t, suffix))
                .collect();
        }

        // Apply duplicate suppression
        tokens = self.apply_duplicate_suppression(tokens);

        // Apply inversion
        tokens = self.invert_tokens(tokens);

        Ok(tokens)
    }

    /// Recursive combination generator
    fn generate_combinations_recursive(
        &self,
        chars: &[char],
        length: usize,
        current: &mut String,
        result: &mut Vec<String>,
    ) -> crate::Result<()> {
        if current.len() == length {
            result.push(current.clone());
            return Ok(());
        }

        for &ch in chars {
            current.push(ch);
            self.generate_combinations_recursive(chars, length, current, result)?;
            current.pop();
        }

        Ok(())
    }

    /// Generate permutations (no repeating characters)
    fn generate_permutations(
        &self,
        chars: &[char],
        length: usize,
        result: &mut Vec<String>,
    ) -> crate::Result<()> {
        if length > chars.len() {
            return Ok(());
        }

        self.permute_helper(
            chars,
            length,
            &mut String::new(),
            &mut std::collections::HashSet::new(),
            result,
        )
    }

    fn permute_helper(
        &self,
        chars: &[char],
        length: usize,
        current: &mut String,
        used: &mut std::collections::HashSet<usize>,
        result: &mut Vec<String>,
    ) -> crate::Result<()> {
        if current.len() == length {
            result.push(current.clone());
            return Ok(());
        }

        for i in 0..chars.len() {
            if !used.contains(&i) {
                current.push(chars[i]);
                used.insert(i);
                self.permute_helper(chars, length, current, used, result)?;
                current.pop();
                used.remove(&i);
            }
        }

        Ok(())
    }

    /// Deduplicate using Blake2b hashing
    pub fn deduplicate(&self, tokens: Vec<String>) -> crate::Result<Vec<String>> {
        let mut seen = HashSet::new();
        let mut unique = Vec::new();

        for token in tokens {
            let hash = self.hash_token(&token);
            if seen.insert(hash) {
                unique.push(token);
            }
        }

        Ok(unique)
    }

    fn hash_token(&self, token: &str) -> u64 {
        use std::collections::hash_map::DefaultHasher;
        let mut hasher = DefaultHasher::new();
        hasher.write(token.as_bytes());
        hasher.finish()
    }

    /// Apply transform pipeline
    pub fn apply_transforms(&self, tokens: Vec<String>) -> crate::Result<Vec<String>> {
        let mut pipeline = crate::transforms::TransformPipeline::new();

        for transform_name in &self.config.transforms {
            let transform = self.parse_transform(transform_name)?;
            pipeline = pipeline.add(transform);
        }

        pipeline.apply_all(tokens)
    }

    /// Parse transform name to Transform enum
    fn parse_transform(&self, name: &str) -> crate::Result<crate::transforms::Transform> {
        match name {
            "leet_basic" => Ok(crate::transforms::Transform::LeetBasic),
            "leet_full" => Ok(crate::transforms::Transform::LeetFull),
            "leet_random" => Ok(crate::transforms::Transform::LeetRandom),
            "reverse" => Ok(crate::transforms::Transform::Reverse),
            "toggle_case" => Ok(crate::transforms::Transform::ToggleCase),
            "upper" => Ok(crate::transforms::Transform::UpperCase),
            "lower" => Ok(crate::transforms::Transform::LowerCase),
            "capitalize" => Ok(crate::transforms::Transform::Capitalize),
            "homoglyph" => Ok(crate::transforms::Transform::Homoglyph),
            "emoji" => Ok(crate::transforms::Transform::EmojiInsertion),
            "keyboard_shift" => Ok(crate::transforms::Transform::KeyboardShift),
            _ => Err(crate::Error::TransformError(format!(
                "Unknown transform: {}",
                name
            ))),
        }
    }

    /// Apply filters
    pub fn apply_filters(&self, tokens: Vec<String>) -> Vec<String> {
        let mut filter_chain = crate::filters::FilterChain::new();

        if let Some(min) = self.config.filters.min_len {
            if let Some(max) = self.config.filters.max_len {
                filter_chain = filter_chain.add_length(min, max);
            }
        }

        if let Some(charset) = &self.config.filters.charset_filter {
            filter_chain = filter_chain.add_charset(charset.clone());
        }

        filter_chain.apply_batch(tokens)
    }

    /// Full pipeline: generate -> dedupe -> transform -> filter
    pub fn generate_full_pipeline(&self) -> crate::Result<Vec<String>> {
        let mut tokens = self.generate_charset()?;

        if self.config.dedupe {
            tokens = self.deduplicate(tokens)?;
        }

        tokens = self.apply_transforms(tokens)?;
        tokens = self.apply_filters(tokens);

        Ok(tokens)
    }

    fn resolve_charset(&self) -> crate::Result<String> {
        if let Some(charset) = &self.config.charset {
            Ok(charset.clone())
        } else {
            Ok(crate::charset::CharsetBuilder::new()
                .add_charset("lower")?
                .build())
        }
    }

    pub fn stats(&self) -> GeneratorStats {
        let state = self.state.lock();
        GeneratorStats {
            tokens_generated: state.tokens_generated,
            unique_tokens: state.dedup_hashes.len(),
        }
    }

    /// Apply duplicate suppression based on config
    fn apply_duplicate_suppression(&self, tokens: Vec<String>) -> Vec<String> {
        if let Some(dup_limit) = &self.config.duplicate_limit {
            tokens.into_iter()
                .filter(|token| self.check_duplicate_limit(token, dup_limit))
                .collect()
        } else {
            tokens
        }
    }

    /// Check if a token violates duplicate character limits
    fn check_duplicate_limit(&self, token: &str, limit_spec: &str) -> bool {
        // Parse limit_spec like "2@" meaning max 2 adjacent @ chars
        // or "@@" meaning no adjacent duplicates for @
        // For simplicity, check for any character repetition > limit
        let limit = if let Some(num_str) = limit_spec.chars().take_while(|c| c.is_numeric()).collect::<String>().parse::<usize>().ok() {
            num_str
        } else {
            1 // Default: no adjacent duplicates
        };

        let chars: Vec<char> = token.chars().collect();
        let mut count = 1;
        
        for i in 1..chars.len() {
            if chars[i] == chars[i-1] {
                count += 1;
                if count > limit {
                    return false;
                }
            } else {
                count = 1;
            }
        }
        
        true
    }

    /// Invert token order (for -i flag, makes first char change most frequently)
    fn invert_tokens(&self, mut tokens: Vec<String>) -> Vec<String> {
        if self.config.invert {
            // Reverse the order - this makes the first character change most frequently
            tokens.reverse();
        }
        tokens
    }

    /// Calculate total combinations for status display
    pub fn calculate_combinations(&self, charset: &str, length: usize) -> u64 {
        let charset_len = charset.chars().count() as u64;
        if self.config.permutations_only {
            // Permutations: n! / (n-r)!
            let mut result = 1u64;
            for i in 0..length {
                result = result.saturating_mul(charset_len.saturating_sub(i as u64));
            }
            result
        } else {
            // Combinations with repetition: n^r
            charset_len.saturating_pow(length as u32)
        }
    }

    /// Show status information before generation
    pub fn show_status_info(&self, charset: &str) -> crate::Result<()> {
        if !self.config.show_status {
            return Ok(());
        }

        println!("📊 Generation Status:");
        println!("  Charset: {} ({} chars)", charset, charset.chars().count());
        println!("  Length range: {} - {}", self.config.min_length, self.config.max_length);
        
        let mut total_combinations = 0u64;
        let mut total_bytes = 0u64;
        
        for len in self.config.min_length..=self.config.max_length {
            let combos = self.calculate_combinations(charset, len);
            total_combinations = total_combinations.saturating_add(combos);
            // Estimate bytes: each token = length + 1 (newline)
            total_bytes = total_bytes.saturating_add(combos.saturating_mul((len + 1) as u64));
        }
        
        println!("  Total combinations: {}", total_combinations);
        println!("  Estimated size: {} bytes ({} KB, {} MB)", 
            total_bytes, 
            total_bytes / 1024,
            total_bytes / 1024 / 1024
        );
        println!();
        
        Ok(())
    }
}


#[derive(Debug, Clone)]
pub struct GeneratorStats {
    pub tokens_generated: u64,
    pub unique_tokens: usize,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generator_creation() {
        let config = crate::Config {
            min_length: 3,
            max_length: 3,
            charset: Some("abc".to_string()),
            ..Default::default()
        };
        let gen = Generator::new(config);
        assert!(gen.is_ok());
    }

    #[test]
    fn test_generate_combinations() {
        let config = crate::Config {
            min_length: 2,
            max_length: 2,
            charset: Some("ab".to_string()),
            ..Default::default()
        };
        let gen = Generator::new(config).unwrap();
        let tokens = gen.generate_charset().unwrap();
        assert_eq!(tokens.len(), 4); // aa, ab, ba, bb
    }

    #[test]
    fn test_deduplication() {
        let config = crate::Config::default();
        let gen = Generator::new(config).unwrap();
        let tokens = vec!["test".to_string(), "test".to_string(), "hello".to_string()];
        let unique = gen.deduplicate(tokens).unwrap();
        assert_eq!(unique.len(), 2);
    }
}
